sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                    #print(matched)
-                    if matched in states:
-                        c = i
-                        while '];' not in lines[c]:
-                            #print(lines[c])
-                            outlines.append(lines[c])
-                            written.append(c)
-                            c += 1
-                        #print(lines[c])
-                        outlines.append(lines[c])
+        for i, line in enumerate(model_lines):
+            node_line = re.match('\\D+(\\d+)\\s\\[\\slabel="\\d.*', line)
+            if node_line:
+                node = int(node_line.group(1))
+                if node in states:
+                    c = i
+                    while '];' not in model_lines[c]:
+                        outlines.append(model_lines[c])
                         written.append(c)
-                    elif matched not in all_states and group == 'ACTIVE_RECON':
-                        if matched != 0:
-                            c = i
-                            while '];' not in lines[c]:
-                                #print(lines[c])
-                                outlines.append(lines[c])
-                                written.append(c)
-                                c += 1
-                            #print(lines[c])
-                            outlines.append(lines[c])
-                            written.append(c)
-                            state_groups['ACTIVE_RECON'].add(matched)
-                        print('ERROR: manually handled', matched, ' in ACTIVE_RECON')
-                # 0 -> 1 [label=
-                '''pattern2 = '\D+(\d+)\s->\s(\d+)\s\[label=.*'
-                SEARCH = re.match(pattern2, line)
-                if SEARCH:
-                    matched = int(SEARCH.group(1))
-                    print(line)
-                    if matched in states:
+                        c += 1
+                    outlines.append(model_lines[c])
+                    written.append(c)
+                elif node not in all_states and group == 'ACTIVE_RECON':
+                    if node != 0:
                         c = i
-                        while '];' not in lines[c]:
-                            #print(lines[c])
-                            outlines.append(lines[c])
+                        while '];' not in model_lines[c]:
+                            outlines.append(model_lines[c])
                             written.append(c)
                             c += 1
-                        #print(lines[c])
-                        outlines.append(lines[c])
+                        outlines.append(model_lines[c])
+                        written.append(c)
+                        state_groups['ACTIVE_RECON'].add(node)
+                    print('ERROR: manually handled', node, ' in ACTIVE_RECON')  # TODO: include edges or not?
+            '''edge_line = re.match('\\D+(\\d+)\\s->\\s(\\d+)\\s\\[label=.*', line)  # 0 -> 1 [label=
```

Comment:
I have kept the part for the edges commented. I have tried running both of them, and I feel like the one without edges is more correct (which is difficult to see since the labels are not really visible due to the size of the S-PDFA). On the other hand, there might be nodes that are only defined in edges. So, I have added a TODO for later...